OpenKeychain/src/test/java/org/sufficientlysecure/keychain/operations/AuthenticationOperationTest.java

```diff
@@ -146,13 +242,107 @@ public void testAuthenticate() throws Exception {
     }
 
     @Test
+    public void testAuthenticateEdDsa() throws Exception {
+
+        byte[] challenge = "dies ist ein challenge ☭".getBytes();
+        byte[] signature;
+
+        KeyRepository keyRepository = KeyRepository.create(RuntimeEnvironment.application);
+
+        long masterKeyId = mStaticRingEdDsa.getMasterKeyId();
+        Long authSubKeyId = keyRepository.getCachedPublicKeyRing(masterKeyId).getSecretAuthenticationId();
+
+        { // sign challenge
+            AuthenticationOperation op = new AuthenticationOperation(RuntimeEnvironment.application,
+                    keyRepository);
+
+            AuthenticationData.Builder authData = AuthenticationData.builder();
+            authData.setAuthenticationMasterKeyId(masterKeyId);
+            authData.setAuthenticationSubKeyId(authSubKeyId);
+            authData.setHashAlgorithm(HashAlgorithmTags.SHA512);
+
+            AuthenticationParcel authenticationParcel = AuthenticationParcel
+                    .createAuthenticationParcel(authData.build(), challenge);
+
+            CryptoInputParcel inputParcel = CryptoInputParcel.createCryptoInputParcel();
+            inputParcel = inputParcel.withPassphrase(mKeyPhrase);
+
+            AuthenticationResult result = op.execute(authData.build(), inputParcel, authenticationParcel);
+
+            Assert.assertTrue("authentication must succeed", result.success());
+
+            signature = result.getSignature();
+        }
+        { // verify signature
+            CanonicalizedPublicKey canonicalizedPublicKey = keyRepository.getCanonicalizedPublicKeyRing(masterKeyId)
+                                                                         .getPublicKey(authSubKeyId);
+            PublicKey publicKey = canonicalizedPublicKey.getJcaPublicKey();
+
+            EdDSAEngine engine = new EdDSAEngine(MessageDigest.getInstance("SHA-512"));
+			engine.initVerify(publicKey);
+			engine.setParameter(EdDSAEngine.ONE_SHOT_MODE);
+			engine.update(challenge);
```

Comment:
Hm, apparently this does not work as expected. Feeding SHA512(challenge) works though. I'm investigating further.

Reply:
EdDSA itself is parametrized with a hashing algorithm, and unlike other (most?) signature schemes can take arbitrary input, not just fixed length. This led to some confusion for me as well, because the interface was inconsistent - JCA assumes signing algorithms are paired with hashing algos, since for a generic interface you have to assume fixed length. A simple eddsa with sha512 parametrization is more equivalent to "RSAwithNONE" than "RSAwithSHA512", at least in openpgp context it was.

Reply:
Yes, and whats makes this worse is that there is a variant HashEdDSA mentioned in [eddsa-20150704.pdf](http://ed25519.cr.yp.to/eddsa-20150704.pdf) (but not in the original paper) that does define a prehash parameter. I was not aware of that until I saw this [issue]( https://github.com/str4d/ed25519-java/issues/34) in ed25519-java for implementing HashEdDSA.